Mask rule

dkiser opened the pull request:
Fixes #2379 

This is an initial attempt at enhancing decision log masks by providing a new structured format for mask rules that allow either erasing or modifying input objects.  Backwards compatibility is maintained with the original mask rule format.

dkiser commented:
This implementation does indeed require the `path` to exist in the `upsert` opcode.  However, if we want to be able to `upsert` a non-existing field in the input, we can modify this to support that easily.

patrick-east commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,234 @@
+	return rule, nil
+}
+
+func (r maskRule) Erase(event *EventV1) {
```
Do we still need to call it `Erase`? Could we make this new one `Mask` or `Apply` or something?

patrick-east commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,234 @@
+	escapedParts := make([]string, len(parts))
+	for i := range parts {
+		var err error
+		unEscapedParts[i], err = url.QueryUnescape(parts[i])
```
Why do we need to keep the unescaped parts?

dkiser replied:
Cleaned this up.

patrick-east commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,234 @@
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				// TODO(dkiser) Do we want to panic instead of return errors here?
+				panic(fmt.Sprintf("illegal mask op value: %s", r.OP))
```
Lets have this return an error up to the logger plugin, the plugin may or may not be able to do anything other than log an error but its better than a panic. That goes for all of the panics in here.

patrick-east commented on plugins/logs/mask_test.go:
```diff
@@ -0,0 +1,482 @@
+
+	for _, tc := range tests {
+		t.Run(tc.note, func(t *testing.T) {
+			// result, err := parsePtr(tc.input)
```
nit: Please remove commented out code

patrick-east commented on plugins/logs/mask_test.go:
```diff
@@ -0,0 +1,482 @@
+	}
+}
+
+func TestMaksRuleErase(t *testing.T) {
```
nit: typo `TestMaksRuleErase` -> `TestMaskRuleErase`

patrick-east commented on plugins/logs/plugin_test.go:
```diff
@@ -483,152 +483,240 @@ func TestPluginReconfigure(t *testing.T) {
+			rawPolicy: []byte(`
+				package system.log
+				mask[{"op": "upsert", "path": "/input/password", "value": x}] {
+					# logic to compute signature value
```
Is this comment supposed to be in the policy snippet?

dkiser commented on docs/content/management.md:
```diff
@@ -426,6 +425,50 @@ There are a few restrictions on the JSON Pointers that OPA will erase:
+#         be mindful of performance when performing complex
+#         opertions in the mask body, such as crypto opertions)
+mask[{"op": "upsert", "path": "/input/password", "value": x}] {
+  input.input.password
```
@patrick-east See last commit update and let me know what you think!

dkiser replied:
See last commit.

dkiser commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,234 @@
+package logs
```
Fixed in last commit.

dkiser replied:
Updated in last commit.

dkiser commented on docs/content/management.md:
```diff
@@ -426,6 +425,50 @@ There are a few restrictions on the JSON Pointers that OPA will erase:
   as undefined. For example `/input/emails/0/value` is allowed but `/input/emails/0` is not.
 
+In order to **modify** the contents of an input field, the **mask** synax has been
+enhanced to support the following optional format.
```
See last commit and let me know what you think.

dkiser commented on docs/content/management.md:
```diff
@@ -426,6 +425,64 @@ There are a few restrictions on the JSON Pointers that OPA will erase:
 * Pointers must refer to object keys. Pointers to array elements will be treated
   as undefined. For example `/input/emails/0/value` is allowed but `/input/emails/0` is not.
 
+In order to **modify** the contents of an input field, the **mask** rule may utilize the following format.
+The structured rule format below provides the ability to choose an opcode for the operation and a value to
+upsert.  If an opcode of **remove** is provided, the value field will be ignored.
```
Sounds good, will work on that when get some minutes.

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+
+	if len(path) == 0 {
+		return rule, fmt.Errorf("mask must be non-empty")
+	} else if path[0] != '/' {
```
nit: use `strings.HasPrefix`

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+				return fmt.Errorf("illegal mask op value: %s", r.OP)
+			}
+		default:
+			return fmt.Errorf("illegal path value value: %s", p)
```
Nit: duplicate `value`

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+
+	if len(r.escapedParts) == 1 {
+		switch p := r.escapedParts[0]; p {
+		case "input":
```
We can probably move the common code below into a function.

dkiser replied:
Refactored the Mask function.

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+				return nil
+			}
+
+			delete(inputObj, fld)
```
Is the delete statement needed if we are overwriting below ?

dkiser replied:
It is not, will remove.

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+
+		default:
+			return nil, fmt.Errorf("invalid mask rule format encountered: %s", reflect.TypeOf(v).String())
+		}
```
Nit:
```golang
return nil, fmt.Errorf("invalid mask rule format encountered: %T", v)
```

ashutosh-narkar commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+	parts := strings.Split(path[1:], "/")
+
+	if parts[0] != "input" && parts[0] != "result" {
+		return rule, fmt.Errorf("mask prefix not allowed")
```
Nit: 
```go
fmt.Errorf("mask prefix not allowed: %v", parts[0]) 
```

ashutosh-narkar commented on plugins/logs/mask_test.go:
```diff
@@ -0,0 +1,489 @@
+			note: "upsert: undefined object: missing key",
+			ptr: maskRule{
+				OP:   maskOPUpsert,
+				Path: "/input/foo/bar/baz",
```
If we had provided a value, would this create an entry in the modified input ?

dkiser replied:
It would not, since the fully qualified json. pointer to `"/input/foo/bar/baz"` does not exist.

tsandall commented on docs/content/management.md:
```diff
@@ -426,6 +425,64 @@ There are a few restrictions on the JSON Pointers that OPA will erase:
+  "input": {
+    "name": "bob",
+    "resource": "user",
+    "password":"**REDACTED**"
```
nit: missing space between "password": and "**REDACTED**"

tsandall commented on plugins/logs/mask.go:
```diff
@@ -0,0 +1,236 @@
+			return fmt.Errorf("illegal path value value: %s", p)
+		}
+
+		parent := r.lookup(r.escapedParts[1:len(r.escapedParts)-1], node)
```
@dkiser the only important comment I have right now is that we should make upsert create intermediate nodes as necessary (like `mkdir -p`). The reason is that otherwise users aren't going to be able to do so reliably because:

* Each mask rule can only generate one operation
* Mask rules are generating a set of operations which are unordered

We could make each make rule generate multiple operations however it'll be somewhat complicated for policies to generate the set of upsert (or other operation types) required to create the hierarchy.

Having the upsert operation create the intermediate nodes resolves these problems.

dkiser replied:
@tsandall  - How should we treat paths with an array index?  Should we ignore these and only allow upsert of objects?

I'll have to mull over how to dynamically created the intermediate nodes.

tsandall replied:
Let's just assume objects for now.

As far as how to create the nodes, here's some code (warning, I haven't tested it):

```golang
func mkdirp(node map[string]interface{}, path []string, value interface{}) bool {
  if len(path) == 0 { return false }
  for i := 0; i < len(path)-1; i++ {
    child, ok := node[path[i]]
    if !ok {
      child := map[string]interface{}{}
      node[path[i]] = child
      node = child
    } else if asobj, ok := child.(map[string]interface{}); ok {
       node = asob
    } else {
      return false
   }
  }
  node[path[len(path)-1]] = value
  return true
}
```

dkiser replied:
Thanks! Playing with these ideas now.

dkiser replied:
@tsandall Added a similar method.  It does make we wonder if we handle `[]interface{}` here as well.

tsandall commented:
@dkiser thanks for adding the upsert support! I tried it out this morning and it works nicely. We should adjust some of the error handling though IMO.

1. The decision logger should validate as much of the mask instructions as possible in one shot and generate an error if any validation issues are found. Let's make the `op` validation happen inside `newMaskRule`. This way if the `op` is missing or incorrect and OPA cannot tell what to do with it, it will be caught.

2. The decision logger should treat undefined paths consistently for remove and upsert operations. Currently if the upsert specifies an undefined path, e.g., a reference into an array, the decision logger reports an error. I think we should suppress these.

Re: (2), I could imagine some users wanting OPA to fail if the path is undefined. I think we can add support for this later if necesarry, either via config or by extending the instructions.

tsandall commented:
@dkiser let us know if you can work on those remaining comments around error handling in the next week or so. If not, we can take on those changes so that this lands in the next release.

dkiser commented:
@tsandall I'll find some time over the next couple of days and see about addressing those last two comments.  Thanks!

dkiser commented:
> @dkiser thanks for adding the upsert support! I tried it out this morning and it works nicely. We should adjust some of the error handling though IMO.
> 
> 1. The decision logger should validate as much of the mask instructions as possible in one shot and generate an error if any validation issues are found. Let's make the `op` validation happen inside `newMaskRule`. This way if the `op` is missing or incorrect and OPA cannot tell what to do with it, it will be caught.
> 2. The decision logger should treat undefined paths consistently for remove and upsert operations. Currently if the upsert specifies an undefined path, e.g., a reference into an array, the decision logger reports an error. I think we should suppress these.
> 
> Re: (2), I could imagine some users wanting OPA to fail if the path is undefined. I think we can add support for this later if necesarry, either via config or by extending the instructions.

@tsandall Re: (2), Do we mean that an upsert to an undefined path for an undefined object type to fail silently?  I think we want to keep the ability to upsert to an undefined path for a defined upsert object type, which right now it just a `map[string]interface{}`?  

Do you have an existing unit test case to point to that behavior should be changed on?

tsandall commented:
@dkiser you're right. The term _undefined_ is not accurate here. I'm referring to this line: https://github.com/open-policy-agent/opa/pull/2433/files#diff-dee023b2dcf1c8c236bac28d4d4fd7b8R189.

In the upsert implementation, if the path refers to a non-object intermediate value, it will result in an error being logged. OTOH, in the remove implementation, if the path refers to a non-object intermediate value, it's treated as undefined and ignored. The behaviour should be consistent. I suggest we remove that error and simply treat the upsert as undefined in this case.

dkiser commented:
@tsandall Can you take a gander at the last commit.  I agree with the configurable option for undefined object path handling (silent or some kind of decision log error) and added low level error handling for this in the mask rule that can be used in a later PR at the `plugin.go` level (left a TODO comment in there).  

Let me know if this new commit addresses your concerns.  Thanks!
